src/Swarm/Game/Log.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- |
+-- Module      :  Swarm.Game.Log
+-- Copyright   :  Ondřej Šebek
+-- Maintainer  :  ondras98@icloud.com
+--
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- A data type to represent in-game logs by robots.
+--
+-- Because of the use of system robots, we sometimes
+-- want to use special kinds of logs that will be
+-- shown to the player.
+--
+-- TODO: Currently we abuse this system for system
+-- logs, which is fun, but we should eventually make
+-- a dedicated `SystemLogEntry` type for 'RuntimeState'
+-- message queue.
+module Swarm.Game.Log (
+  LogSource (..),
+  ErrorLevel (..),
+
+  -- * Robot log entries
+  LogEntry (..),
+  leText,
+  leSaid,
+  leRobotName,
+  leTime,
+  leLocation,
+  leRobotID,
+) where
+
+import Control.Lens hiding (contains)
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Text (Text)
+import GHC.Generics (Generic)
+import Swarm.Util.Location (Location)
+
+-- | Severity of the error - critical errors are bugs
+--   and should be reported as Issues.
+data ErrorLevel = Debug | Warning | Error | Critical
+  deriving (Show, Eq, Ord, Generic, FromJSON, ToJSON)
+
+-- | Source of the robot log.
+data LogSource
+  = -- | Log produced by 'Swarm.Language.Syntax.Say'
+    Said
+  | -- | Log produced by 'Swarm.Language.Syntax.Log'
+    Logged
+  | -- | Log produced by an exception or system.
+    ErrorTrace ErrorLevel
+  deriving (Show, Eq, Ord, Generic, FromJSON, ToJSON)
+
+-- | An entry in a robot's log.
+data LogEntry = LogEntry
+  { _leTime :: Integer
+  -- ^ The time at which the entry was created.
+  --   Note that this is the first field we sort on.
+  , _leSaid :: LogSource
+  -- ^ Whether this log records a said message.
+  , _leRobotName :: Text
+  -- ^ The name of the robot that generated the entry.
+  , _leRobotID :: Int
+  -- ^ The ID of the robot that generated the entry.
+  , _leLocation :: Location
+  -- ^ Location of the robot at log entry creation.
+  , _leText :: Text
```

Comment:
At https://github.com/swarm-game/swarm/pull/959#issuecomment-1372626765 we discussed maybe storing something more structured here than just a `Text`, which I think is worth exploring.  But we should probably leave that to a different PR.

Reply:
I would like to split a separate type for system logs which should also use structured errors.

I think that will make the design clear, now it's all tangled and adding anything to it would be a mistake IMO.